File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* tail = dummy;
        
        while (true){
            ListNode* Kth = getKthNode(tail, k);
            if (!Kth) break;

            ListNode* nextGroup = Kth->next;
            ListNode* curr = tail->next;
            ListNode* prev = Kth->next;

            while(curr != nextGroup){
                ListNode* temp = curr->next;
                curr->next = prev;
                prev = curr;
                curr = temp;
            }

            ListNode* temp = tail->next;
            tail->next = prev;
            tail = temp;
        }

        return dummy->next;
    }

    ListNode* getKthNode(ListNode* Node, int k){
        while(Node && k){
            Node = Node->next;
            k -= 1;
        }
        return Node;
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        tail = dummy

        while True:
            kth = self.getKthNode(tail, k)
            
            if not kth: break
            groupNext = kth.next
            prev = kth.next
            curr = tail.next
            while curr != groupNext:
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp
            
            temp = tail.next
            tail.next = kth
            tail = temp
        return dummy.next
    
    def getKthNode(self, node: Optional[ListNode], k: int) -> Optional[ListNode]:

        while node and k>0:
            k -= 1
            node = node.next

        return node 